### PR TITLE
Python bindings for crop_sicd

### PR DESCRIPTION
**Pull request by slipschutz**
This merge adds swig wrapping for the crop sicd utility.  The c++ implementation was updated to allow for just the metadata object to be cropped.

### CODE REVIEW

File: six/modules/c++/six.sicd/include/six/sicd/CropUtils.h [hunk 1]
```diff
@@ -116,6 +116,21 @@ void cropSICD(six::NITFReadControl& reader,
               const std::vector<scene::LatLonAlt>& corners,
               const std::string& outPathname,
               bool trimCornersIfNeeded = true);
+
+/*
+ * Given a six::Sicd::ComplexData object and a cropping region,
+ * update the meta data in the same way as the functions above.
```
Comment (asylvest) [inline six/modules/c++/six.sicd/include/six/sicd/CropUtils.h]: `meta data` --> `metadata`

File: six/modules/c++/six.sicd/include/six/sicd/CropUtils.h [hunk 2]
```diff
@@ -116,6 +116,21 @@ void cropSICD(six::NITFReadControl& reader,
...
+ * \param complexData the input complexData object
+ * \param aoiOffset Upper left corner of AOI
+ * \param aoiDims Size of AOI
+ *
+ * \returns A cloned complex data object where the 
+ * meta data has been updated to refelect croped aoi.
+ */
+six::sicd::ComplexData* const
+cropMetaData(const six::sicd::ComplexData* complexData,
```
Comment (asylvest) [inline six/modules/c++/six.sicd/include/six/sicd/CropUtils.h]: Pass in by const ref rather than const pointer

File: six/modules/c++/six.sicd/include/six/sicd/CropUtils.h [hunk 3]
```diff
@@ -116,6 +116,21 @@ void cropSICD(six::NITFReadControl& reader,
...
+ * Given a six::Sicd::ComplexData object and a cropping region,
+ * update the meta data in the same way as the functions above.
+ * \param complexData the input complexData object
+ * \param aoiOffset Upper left corner of AOI
+ * \param aoiDims Size of AOI
+ *
+ * \returns A cloned complex data object where the 
+ * meta data has been updated to refelect croped aoi.
+ */
+six::sicd::ComplexData* const
```
Comment (asylvest) [inline six/modules/c++/six.sicd/include/six/sicd/CropUtils.h]: Return by `auto_ptr` rather than raw pointer.  Swig should be able to deal with this.

File: six/modules/c++/six.sicd/include/six/sicd/CropUtils.h (hunk 2 above)
Comment (asylvest) [inline six/modules/c++/six.sicd/include/six/sicd/CropUtils.h]: `cropMetaData` --> `cropMetadata()`

File: six/modules/python/six.sicd/source/six_sicd.i [hunk 4]
```diff
@@ -124,6 +124,28 @@ Data* readNITF(const std::string& pathname,
...
 }
 
+
+void crop_SICD(const std::string& inPathname,
+               const std::vector<std::string>& schemaPaths,
+               const types::RowCol<size_t>& aoiOffset,
+               const types::RowCol<size_t>& aoiDims,
+               const std::string& outPathname);
+              
+void crop_SICD(const std::string& inPathname,
```
Comment (asylvest) [inline six/modules/python/six.sicd/source/six_sicd.i]: `crop_SICD()` --> `cropSICD()`

File: six/modules/python/six.sicd/source/six_sicd.i [hunk 5]
```diff
@@ -124,6 +124,28 @@ Data* readNITF(const std::string& pathname,
     return reinterpret_cast<Data*>(reader.interleaved(region, 0));
 }
 
+
+void crop_SICD(const std::string& inPathname,
```
Comment (asylvest) [inline six/modules/python/six.sicd/source/six_sicd.i]: Why is this here?  Doesn't seem to do anything..?

File: six/modules/python/six.sicd/source/six_sicd.i [hunk 6]
```diff
@@ -199,6 +221,17 @@ void writeNITF(const std::string& pathname, const std::vector<std::string>&
 Data* readNITF(const std::string& pathname,
         const std::vector<std::string>& schemaPaths);
 
+void crop_SICD(const std::string& inPathname,
```
Comment (asylvest) [inline six/modules/python/six.sicd/source/six_sicd.i]: Why is this here?  Doesn't seem to do anything..?

File: six/modules/python/six.sicd/source/six_sicd.i [hunk 7]
```diff
@@ -199,6 +221,17 @@ void writeNITF(const std::string& pathname, const std::vector<std::string>&
...
+               const std::vector<std::string>& schemaPaths,
+               const types::RowCol<size_t>& aoiOffset,
+               const types::RowCol<size_t>& aoiDims,
+               const std::string& outPathname);
+
+six::sicd::ComplexData* const cropMetaData(
```
Comment (asylvest) [inline six/modules/python/six.sicd/source/six_sicd.i]: Why is this here?  Doesn't seem to do anything..?

Is this to get Swig to generate the function?  If you do a `#include <six/sicd/CropUtils.h>` on top and a `%include <six/sicd/CropUtils.h>` farther down, I don't think you need these... Swig should just do it.

@JonathanMeans can help if you're having trouble getting Swig to generate what you want.

File: six/modules/python/six.sicd/source/six_sicd.i (hunk 6 above)
Comment (slipschutz) [reply six/modules/python/six.sicd/source/six_sicd.i]: I can't directly call cropSICD from CropUtils.h because this line needs to be called first.
```
six::XMLControlFactory::getInstance().addCreator(
          six::DataType::COMPLEX,
	  new six::XMLControlCreatorT<six::sicd::ComplexXMLControl>());
```

So I made a wrapper called crop_SICD to call that line first.  This line is to tell swig to wrap that function.

File: six/modules/python/six.sicd/source/six_sicd.i (hunk 7 above)
Comment (slipschutz) [reply six/modules/python/six.sicd/source/six_sicd.i]: As mentioned in comment above I don't want swig to wrap all the version of cropSICD in CropUtils.h since those can't be called directly.

File: six/modules/python/six.sicd/source/six_sicd.i (hunk 5 above)
Comment (slipschutz) [reply six/modules/python/six.sicd/source/six_sicd.i]: Lines 128-147 are defining a wrapper on cropSICD that first calls 
'''
 six::XMLControlFactory::getInstance().addCreator(
          six::DataType::COMPLEX,
	  new six::XMLControlCreatorT<six::sicd::ComplexXMLControl>());
'''

File: six/modules/python/six.sicd/source/six_sicd.i (hunk 4 above)
Comment (slipschutz) [reply six/modules/python/six.sicd/source/six_sicd.i]: I called it crop_SICD because there is already a function in CropUtils.h with the name cropSICD and the same arguments (the one I am wrapping).  Is there a better way to get around this?

File: six/modules/c++/six.sicd/include/six/sicd/CropUtils.h (hunk 2 above)
Comment (slipschutz) [reply six/modules/c++/six.sicd/include/six/sicd/CropUtils.h]: I don't understand this comment

File: six/modules/python/six.sicd/source/six_sicd.i (hunk 4 above)
Comment (asylvest) [reply six/modules/python/six.sicd/source/six_sicd.i]: @JonathanMeans can you show him how to do this?  There's a way with (I think it's) `%ignore` that you can tell Swig to ignore certain overloadings but still let some through... that way we don't need the underscore in these.

Comment (JonathanMeans) [reply six/modules/python/six.sicd/source/six_sicd.i]: Try adding `%rename(_cropSICDImpl) six::sicd::cropSICD;`

Comment (slipschutz) [reply six/modules/python/six.sicd/source/six_sicd.i]: I tried renaming and ignoring, but it didn't work.  I think the issue is that the CropUtils.h is being included in the generated six_sicd.swigwrap_2.cxx file and I am trying to define a new c++ function with the same signature as one in CropUtils.h.  The error is happening when I gcc is compiling six_sicd.swigwrap_2.cxx:
```
six/modules/python/six.sicd/source/six_sicd.swigwrap_2.cxx:3566:25: error: call of overloaded ‘cropSICD(const string&, const std::vector<std::basic_string<char> >&, const types::RowCol<long unsigned int>&, const types::RowCol<long unsigned int>&, const string&)’ is ambiguous
     aoiDims, outPathname);
```
The only reason I have this function is call six::XMLControlFactory::getInstance().addCreator(),

**Comment by slipschutz** [comment]
@asylvest I have made the changes you requested.

**Comment by slipschutz** [comment]
@asylvest Did you see the push I did with the updates requested?

File: six/modules/python/six.sicd/tests/test_crop_sicd.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python
+
+#
+# =========================================================================
+# This file is part of six.sicd-python
+# =========================================================================
+#
+# (C) Copyright 2004 - 2015, MDA Information Systems LLC
```
Comment (asylvest) [inline six/modules/python/six.sicd/tests/test_crop_sicd.py]: Update to 2019

**Comment by slipschutz** [comment]
@asylvest The change has been done.
